### DISCUSSION TITLE
low-level offboard controller

### DISCUSSION THREAD

**Discussion by Finn-Sueberkrueb** [bitcraze/discussions#413] [Q&A]
Hej Bitcraze-community,

I plan to use the Crazyflie platform for experiments with low-level offboard controllers. (I want to send PWM signals directly via the Crazyradio.)

1. Are there measurements how long a round trip from Cflib via Crazyradio to Crazyfly and back takes for a message?
2. Are there any more recent experiments with low-level offboard controllers apart from https://forum.bitcraze.io/viewtopic.php?t=2730?
3. The small drone is probably very agile and needs a high control frequency. Is there any experience how slow the Attitude Rate PID (implemented with 500 Hz) can be in order for the drone to still hover?

**Comment by ataffanel** [comment]
Hi, this is not an easy task, I have usually advice to run external controller at least on step higher: sending attitude rate or absolute attitude to the Crazyflie to make it less communication-latency dependent.

1. The round-trip latency is from 3.5 to 5ms (smaller packet induces less latency, 5ms is for 28B packets) when using the Crazyflie echo port: it is a low level port that repeats back what it received so it allows to test radio latency at the lowest level.
2. I am not aware or more recent experiments
3. We have run the internal low level controller at 500Hz without problem. We feel that it should work down to 200Hz but this has not been tested.

My feeling is that the best way would be to implement your own radio packets using the app channel port, bypassing the logging subsystem. This way you will increase your chance of having very low latency and you will be able to send back IMU measurements synchronously (the log subsystem is asynchronous).

**Comment by Finn-Sueberkrueb** [comment]
thank you very much for the helpful input!

I am now considering whether I can establish a connection via the ESP32 on the AI deck with lower latency. Inspired by this link: [https://hackaday.io/project/161896-linux-espnow](https://hackaday.io/project/161896-linux-espnow)

Are there any measurements of the latency between the ESP32 on the AI deck and the Crazyflie controller?

**Reply by ataffanel** [reply]
I do not think you will get lower latency though the AI deck. We do not have any measurement but the link between the esp32 and the stm32 is designed for safety and bandwidth not for low latenty (there is a brunch of queues in the way and packets are hold until a big enough packet can be transferred over the UART). So you likely will have to write some special esp32 code for low latency.

By the way this is a very interesting link. We always have had quite bad experience with Wifi when it comes to latency and random jitter. It is interesting to see that it might be possible to have a more reliable real time link over wifi.

**Reply by ntukenmez3** [reply]
Hi all, at that point, I have a question about off-board controller design for a crazyflie. Is it possible to design such that controller via Python, I mean we'd like to implement our modified LQR controlller on Crazyflie. Could you share a useful guide to make it possible, thanks.

**Reply by krichardsson** [reply]
You can either run your controller in your computer and use python, or you can run it in the Crazyflie but then you have to write it in C.
We don't have any example code for offboard controllers (as far as I know), but there is some docs for [adding onboard controllers](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/oot/#oot-controllers)

**Reply by ntukenmez3** [reply]
I got it and I checked the page you mentioned two weeks ago.
